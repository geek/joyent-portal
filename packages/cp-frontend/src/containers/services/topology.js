import React, { Component } from 'react';
import { compose, graphql } from 'react-apollo';
import { connect } from 'react-redux';
import styled from 'styled-components';
import forceArray from 'force-array';
import unitcalc from 'unitcalc';

import ServicesQuery from '@graphql/Services.gql';
import { processServicesForTopology } from '@root/state/selectors';
import { toggleServicesQuickActions } from '@root/state/actions';
import { withNotFound, GqlPaths } from '@containers/navigation';
import { LayoutContainer } from '@components/layout';
import { Loader, ErrorMessage } from '@components/messaging';
import { Topology } from 'joyent-ui-toolkit';

const StyledBackground = styled.div`
  padding: ${unitcalc(4)};
  background-color: ${props => props.theme.whiteActive};
`;

const StyledContainer = styled.div`
  position: relative;
`;

export class ServicesTopology extends Component {
  constructor(props) {
    super(props);

    this.state = {
      errors: {}
    };
  }

  ref(name) {
    this._refs = this._refs || {};

    return el => {
      this._refs[name] = el;
    };
  }

  render() {
    const {
      url,
      push,
      deploymentGroup,
      services,
      loading,
      error,
      toggleServicesQuickActions,
    } = this.props;

    if (loading && !forceArray(services).length) {
      return (
        <LayoutContainer center>
          <Loader />
        </LayoutContainer>
      );
    }

    if (error) {
      return (
        <LayoutContainer>
          <ErrorMessage
            title="Ooops!"
            message="An error occurred while loading your services."
          />
        </LayoutContainer>
      );
    }

    if (
      deploymentGroup &&
      deploymentGroup.status === 'PROVISIONING' &&
      !forceArray(services).length
    ) {
      return (
        <LayoutContainer center>
          <Loader msg="Just a moment, weâ€™re on it" />
        </LayoutContainer>
      );
    }

    const handleQuickActionsClick = (evt, tooltipData) => {
      const container = this._refs.container;
      const containerRect = container.getBoundingClientRect();
      const position = {
        top: `${containerRect.top + window.scrollY + tooltipData.position.top}px`,
        left: `${containerRect.left + window.scrollX + tooltipData.position.left}px`
      }
      const data = {
        ...tooltipData,
        position
      }
      toggleServicesQuickActions(data);
    };

    const handleNodeTitleClick = (evt, { service }) => {
      push(`${url.split('/').slice(0, 3).join('/')}/services/${service.slug}`);
    };

    let renderedError = null;

    if (
      this.state.errors.stop ||
      this.state.errors.start ||
      this.state.errors.restart
    ) {
      const message = this.state.errors.stop
        ? 'An error occurred while attempting to stop your service.'
        : this.state.errors.start
          ? 'An error occurred while attempting to start your service.'
          : this.state.errors.restart
            ? 'An error occurred while attempting to restart your service.'
            : '';

      renderedError = (
        <LayoutContainer>
          <ErrorMessage title="Ooops!" message={message} />
        </LayoutContainer>
      );
    }

    return (
      <div>
        {renderedError}
        <StyledBackground>
          <StyledContainer>
            <div ref={this.ref('container')}>
              <Topology
                services={services}
                onQuickActionsClick={handleQuickActionsClick}
                onNodeTitleClick={handleNodeTitleClick}
              />
            </div>
          </StyledContainer>
        </StyledBackground>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  url: ownProps.match.url.replace(/\/$/, ''),
  push: ownProps.history.push
});

const mapDispatchToProps = dispatch => ({
  toggleServicesQuickActions: data => dispatch(toggleServicesQuickActions(data))
});

const UiConnect = connect(mapStateToProps, mapDispatchToProps);

const ServicesGql = graphql(ServicesQuery, {
  options(props) {
    return {
      pollInterval: 1000,
      variables: {
        deploymentGroupSlug: props.match.params.deploymentGroup
      }
    };
  },
  props: ({ data: { deploymentGroup = {}, loading, error } }) => ({
    deploymentGroup,
    services: deploymentGroup
      ? processServicesForTopology(deploymentGroup.services)
      : null,
    loading,
    error
  })
});

const ServicesTopologyWithData = compose(
  ServicesGql,
  UiConnect,
  withNotFound([ GqlPaths.DEPLOYMENT_GROUP ])
)(ServicesTopology);

export default ServicesTopologyWithData;
