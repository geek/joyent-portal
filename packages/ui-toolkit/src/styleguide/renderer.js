import React from 'react';
import styled, { ThemeProvider } from 'styled-components';
import remcalc from 'remcalc';
import is from 'styled-is';

import { Grid, Row, Col } from '../grid';
import theme from '../theme';

const Main = styled(Row)`
  padding-top: ${remcalc(40)};
`;

const Sticky = styled.div`
  position: sticky;
  top: 0;
  z-index: 9;

  ${is('sidebar')`
    top: ${remcalc(100)};
    max-height: 85vh;
    overflow: auto;
  `};
`;

const fullTheme = {
  ...theme,
  spacing: {
    0.5: remcalc(4),
    0: remcalc(0),
    1: remcalc(6),
    2: remcalc(12),
    3: remcalc(18),
    4: remcalc(24),
    5: remcalc(30),
    6: remcalc(36),
    7: remcalc(42),
    8: remcalc(48),
    9: remcalc(54),
    10: remcalc(60)
  }
};

const StyleGuideRenderer = ({
  title,
  homepageUrl,
  children,
  toc,
  hasSidebar
}) => (
  <ThemeProvider theme={fullTheme}>
    <Grid>
      <Main>
        {hasSidebar && (
          <Col xs={3}>
            <Sticky sidebar>{toc}</Sticky>
          </Col>
        )}
        <Col xs={hasSidebar ? 9 : 12} lg={hasSidebar ? 8 : 12}>
          {children}
        </Col>
      </Main>
    </Grid>
  </ThemeProvider>
);

export default StyleGuideRenderer;
