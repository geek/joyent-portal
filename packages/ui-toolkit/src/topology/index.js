import React from 'react';
import { Svg } from 'normalized-styled-components';
import PropTypes from 'prop-types';

import Baseline from '../baseline';
import Constants from './constants';
import { createSimulation } from './simulation';
import TopologyNode from './node';
import TopologyLink from './link';
import TopologyLinkArrow from './link/arrow';
import { calculateLineLayout } from './link/functions';

const StyledSvg = Svg.extend`
  width: 100%;
  height: 1400px;
`;

/**
 * @example ./usage.md
 */
class Topology extends React.Component {
  componentWillMount() {
    const services = this.getServicesWithoutConsul();
    const svgSize = this.getSvgSize();

    const { nodes, links, simulation } = createSimulation(services, svgSize);

    this.setState({
      nodes,
      links,
      simulation
    });
  }

  getServicesWithoutConsul() {
    return this.props.services.reduce((acc, service, index) => {
      if (service.id !== 'consul') acc.push(service);
      return acc;
    }, []);
  }

  getSvgSize() {
    if (document.getElementById('topology-svg')) {
      return document.getElementById('topology-svg').getBoundingClientRect();
    }

    const windowWidth =
      window.innerWidth ||
      document.documentElement.clientWidth ||
      document.body.clientWidth;

    return {
      width: windowWidth - 2 * 24,
      height: 1400
    };
  }

  constrainNodePosition(x, y, children = false) {
    const svgSize = this.getSvgSize();

    const nodeRect = children
      ? Constants.nodeRectWithChildren
      : Constants.nodeRect;

    if (x < nodeRect.right + 2) {
      x = nodeRect.right + 2;
    } else if (x > svgSize.width + nodeRect.left - 2) {
      x = svgSize.width + nodeRect.left - 2;
    }

    if (y < -nodeRect.top + 2) {
      y = -nodeRect.top + 2;
    } else if (y > svgSize.height - nodeRect.bottom - 2) {
      y = svgSize.height - nodeRect.bottom - 2;
    }

    return {
      x,
      y
    };
  }

  findNode(nodeId) {
    return this.state.nodes.reduce(
      (acc, simNode, index) => (simNode.id === nodeId ? simNode : acc),
      {}
    );
  }

  getConsulNodePosition() {
    const svgSize = this.getSvgSize();
    const x = svgSize.width - Constants.nodeSize.width;

    return {
      x,
      y: 0
    };
  }

  getConstrainedNodePosition(nodeId, children = false) {
    const node = this.findNode(nodeId);
    return this.constrainNodePosition(node.x, node.y, children);
  }

  findNodeData(nodesData, nodeId) {
    return nodesData.reduce(
      (acc, nodeData, index) => (nodeData.id === nodeId ? nodeData : acc),
      {}
    );
  }

  setDragInfo(dragging, nodeId = null, position = {}) {
    this.dragInfo = {
      dragging,
      nodeId,
      position
    };
  }

  render() {
    const { onQuickActionsClick, onNodeTitleClick, services } = this.props;

    const { nodes, links } = this.state;

    const nodesData = services.map((service, index) => {
      const nodePosition = service.id === 'consul'
        ? this.getConsulNodePosition()
        : this.getConstrainedNodePosition(service.id, service.children);

      return {
        ...service,
        ...nodePosition
      };
    });

    // TODO links will need to know whether a service has children
    // if it does, the height of it will be different
    const linksData = links
      .map((link, index) => ({
        source: this.findNodeData(nodesData, link.source.id),
        target: this.findNodeData(nodesData, link.target.id)
      }))
      .map((linkData, index) => calculateLineLayout(linkData, index));

    const onDragStart = (evt, nodeId) => {
      // It's this node's position that we'll need to update

      const x = evt.changedTouches ? evt.changedTouches[0].pageX : evt.clientX;
      const y = evt.changedTouches ? evt.changedTouches[0].pageY : evt.clientY;

      this.setDragInfo(true, nodeId, {
        x,
        y
      });
    };

    const onDragMove = evt => {
      if (this.dragInfo && this.dragInfo.dragging) {
        const x = evt.changedTouches
          ? evt.changedTouches[0].pageX
          : evt.clientX;
        const y = evt.changedTouches
          ? evt.changedTouches[0].pageY
          : evt.clientY;

        const offset = {
          x: x - this.dragInfo.position.x,
          y: y - this.dragInfo.position.y
        };

        const dragNodes = nodes.map((simNode, index) => {
          if (simNode.id === this.dragInfo.nodeId) {
            return {
              ...simNode,
              x: simNode.x + offset.x,
              y: simNode.y + offset.y
            };
          }
          return {
            ...simNode
          };
        });

        this.setState({
          nodes: dragNodes
        });

        this.setDragInfo(true, this.dragInfo.nodeId, {
          x,
          y
        });
      }
    };

    const onDragEnd = evt => {
      this.setDragInfo(false);
    };

    const renderedNode = (n, index) =>
      <TopologyNode
        key={index}
        data={n}
        index={index}
        onDragStart={onDragStart}
        onNodeTitleClick={onNodeTitleClick}
        onQuickActions={onQuickActionsClick}
        connected={n.id !== 'consul'}
      />;

    const renderedLink = (l, index) =>
      <TopologyLink key={index} data={l} index={index} />;

    const renderedLinkArrow = (l, index) =>
      <TopologyLinkArrow key={index} data={l} index={index} />;

    const renderedNodes = this.dragInfo && this.dragInfo.dragging
      ? nodesData
          .filter((n, index) => n.id !== this.dragInfo.nodeId)
          .map((n, index) => renderedNode(n, index))
      : nodesData.map((n, index) => renderedNode(n, index));

    const renderedLinks = linksData.map((l, index) => renderedLink(l, index));

    const renderedLinkArrows = this.dragInfo && this.dragInfo.dragging
      ? linksData
          .filter((l, index) => l.target.id !== this.dragInfo.nodeId)
          .map((l, index) => renderedLinkArrow(l, index))
      : linksData.map((l, index) => renderedLinkArrow(l, index));

    const dragNode = !this.dragInfo || !this.dragInfo.dragging
      ? null
      : renderedNode(
          nodesData.reduce((dragNode, n, index) => {
            if (n.id === this.dragInfo.nodeId) {
              return n;
            }
            return dragNode;
          }, {})
        );

    const dragLinkArrow = !this.dragInfo ||
      !this.dragInfo.dragging ||
      renderedLinkArrows.length === renderedLinks.length
      ? null
      : renderedLinkArrow(
          linksData.reduce((dragLinkArrow, l, index) => {
            if (l.target.id === this.dragInfo.nodeId) {
              return l;
            }
            return dragLinkArrow;
          }, {})
        );

    return (
      <StyledSvg
        onMouseMove={onDragMove}
        onTouchMove={onDragMove}
        onMouseUp={onDragEnd}
        onTouchEnd={onDragEnd}
        onTouchCancel={onDragEnd}
        id="topology-svg"
      >
        <g>
          {renderedNodes}
        </g>
        <g>
          {renderedLinks}
        </g>
        <g>
          {renderedLinkArrows}
        </g>
        <g>
          {dragNode}
        </g>
        <g>
          {dragLinkArrow}
        </g>
      </StyledSvg>
    );
  }
}

Topology.propTypes = {
  onQuickActionsClick: PropTypes.func,
  onNodeTitleClick: PropTypes.func,
  services: PropTypes.array
};

export default Baseline(Topology);

export { default as TopologyNode } from './node';
export { default as TopologyLink } from './link';
