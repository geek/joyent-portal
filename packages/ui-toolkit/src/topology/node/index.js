import React from 'react';
import PropTypes from 'prop-types';
import Constants from '../constants';
import GraphNodeTitle from './title';
import GraphNodeButton from './button';
import GraphNodeContent from './content';
import { GraphNodeRect, GraphShadowRect } from './shapes';
import Baseline from '../../baseline';

const GraphNode = ({
  connected,
  data,
  index,
  onDragStart,
  onNodeTitleClick,
  onQuickActions
}) => {
  const { width, height } = data.children
    ? Constants.nodeSizeWithChildren
    : Constants.nodeSize;

  const { left, top } = data.children
    ? Constants.nodeRectWithChildren
    : Constants.nodeRect;

  let x = data.x;
  let y = data.y;
  if (connected) {
    x = data.x + left;
    y = data.y + top;
  }

  const onButtonClick = evt => {
    const tooltipPosition = {
      x: data.x + Constants.buttonRect.x + Constants.buttonRect.width / 2,
      y: data.y + Constants.buttonRect.y + Constants.buttonRect.height
    };

    if (connected) {
      tooltipPosition.x += left;
      tooltipPosition.y += top;
    }

    const d = {
      service: data,
      position: {
        left: tooltipPosition.x,
        top: tooltipPosition.y
      }
    };

    onQuickActions(evt, d);
  };

  const onTitleClick = evt => onNodeTitleClick(evt, { service: data });

  const onStart = evt => {
    evt.preventDefault();
    onDragStart(evt, data.id);
  };

  const nodeRectEvents = connected
    ? {
        onMouseDown: onStart,
        onTouchStart: onStart
      }
    : {};

  const nodeContent = data.children
    ? data.children.map((d, i) =>
        <GraphNodeContent
          key={i}
          child
          connected={connected}
          data={d}
          index={i}
        />
      )
    : <GraphNodeContent connected={connected} data={data} />;

  return (
    <g transform={`translate(${x}, ${y})`}>
      <GraphShadowRect
        x={0}
        y={3}
        width={width}
        height={height}
        connected={connected}
      />
      <GraphNodeRect
        x={0}
        y={0}
        width={width}
        height={height}
        connected={connected}
        {...nodeRectEvents}
      />
      <GraphNodeTitle
        connected={connected}
        data={data}
        onNodeTitleClick={onTitleClick}
      />
      <GraphNodeButton
        connected={connected}
        index={index}
        onButtonClick={onButtonClick}
      />
      {nodeContent}
    </g>
  );
};

GraphNode.propTypes = {
  connected: PropTypes.bool,
  data: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  onDragStart: PropTypes.func,
  onNodeTitleClick: PropTypes.func,
  onQuickActions: PropTypes.func
};

export default Baseline(GraphNode);
