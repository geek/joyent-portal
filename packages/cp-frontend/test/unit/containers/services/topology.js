
/**
 * @jest-environment jsdom
 */

import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { ServicesTopology } from '@containers/services/topology.js';
import { Router, Store, services } from '../../mocks';

it('renders <ServicesTopology /> without throwing', () => {
  const tree = renderer
    .create(
      <Store>
        <Router>
          <ServicesTopology
            services={services}
          />
        </Router>
      </Store>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
