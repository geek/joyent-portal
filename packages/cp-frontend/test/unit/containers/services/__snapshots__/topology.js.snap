// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <ServicesTopology /> without throwing 1`] = `
.c5 {
  stroke: ;
  stroke-width: 1.5;
  stroke: ;
}

.c3 {
  stroke: ;
  fill: ;
  stroke-width: 1.5;
  rx: 4;
  ry: 4;
  stroke: ;
  fill: ;
}

.c4 {
  font-family: "Libre Franklin",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica,sans-serif;
  font-weight: 400;
  fill: ;
  font-size: 16px;
  font-weight: 600;
  fill: ;
  cursor: pointer;
}

.c9 {
  font-family: "Libre Franklin",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,Helvetica,sans-serif;
  font-weight: 400;
  fill: ;
  font-size: 12px;
  opacity: 0.8;
  fill: ;
}

.c7 {
  opacity: 0;
  cursor: pointer;
}

.c7:focus {
  outline: none;
}

.c6 {
  fill: ;
  fill: ;
}

.c8 {
  fill: ;
  fill: ;
}

.c2 {
  width: 100%;
  height: 1000px;
}

.c2:not(:root) {
  overflow: hidden;
}

.c0 {
  padding: 1.5rem;
  background-color: ;
}

.c1 {
  position: relative;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <div>
        <svg
          className="c2"
          id="topology-svg"
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
        >
          <g>
            <g
              transform="translate(0, 0)"
            >
              <rect
                className="c3"
                height={114}
                width={180}
                x={0}
                y={0}
              />
              <g>
                <text
                  className="c4"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  x={12}
                  y={30}
                >
                  Nginx
                </text>
              </g>
              <g
                transform="translate(140, 0)"
              >
                <line
                  className="c5"
                  x1={0}
                  x2={0}
                  y1={0}
                  y2={48}
                />
                <circle
                  className="c6"
                  cx={20}
                  cy={18}
                  r={2}
                />
                <circle
                  className="c6"
                  cx={20}
                  cy={24}
                  r={2}
                />
                <circle
                  className="c6"
                  cx={20}
                  cy={30}
                  r={2}
                />
                <rect
                  className="c7"
                  height={48}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  tabIndex={100}
                  width={40}
                />
              </g>
              <g
                transform="translate(0, 48)"
              >
                <line
                  className="c5"
                  x1={0}
                  x2={180}
                  y1={0}
                  y2={0}
                />
                <g
                  transform="translate(12, 11)"
                >
                  <g
                    transform="translate(0, 0)"
                  >
                    <svg
                      className=""
                      height="18"
                      viewBox="0 0 18 18"
                      width="18"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <title>
                        icon: error
                      </title>
                      <g
                        fill="none"
                        fillRule="evenodd"
                      >
                        <circle
                          cx="9"
                          cy="9"
                          fill="#DA4B42"
                          r="9"
                        />
                        <g
                          fill="#FFF"
                        >
                          <path
                            d="M11.121 5.464l1.415 1.415-5.657 5.657-1.415-1.415z"
                          />
                          <path
                            d="M12.536 11.121l-1.415 1.415-5.657-5.657L6.88 5.464z"
                          />
                        </g>
                      </g>
                    </svg>
                  </g>
                  <g
                    transform="translate(30, 4.5)"
                  >
                    <svg
                      active={undefined}
                      className="c8"
                      consul={undefined}
                      height="9"
                      viewBox="0 0 18 9"
                      width="18"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <title>
                        icon: instances
                      </title>
                      <path
                        d="M4.5 0a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9m8.59 0c-.375 0-.739.047-1.09.12 2.184.456 3.818 2.239 3.818 4.38S14.184 8.424 12 8.88c.351.073.715.12 1.09.12C15.803 9 18 6.985 18 4.5S15.802 0 13.09 0M14 4.5C14 6.985 11.802 9 9.09 9c-.375 0-.739-.047-1.09-.12 2.184-.456 3.818-2.239 3.818-4.38S10.184.576 8 .12C8.351.047 8.715 0 9.09 0 11.803 0 14 2.015 14 4.5"
                        fillRule="evenodd"
                      />
                    </svg>
                  </g>
                  <text
                    className="c9"
                    x={54}
                    y={14}
                  >
                    1 inst.
                  </text>
                  <g
                    transform="translate(54, 36)"
                  >
                    <text
                      className="c9"
                    >
                      1
                                  running
                    </text>
                  </g>
                </g>
              </g>
            </g>
          </g>
          <g />
          <g />
          <g />
          <g />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
